File: lib/guidance/pure_proportional_navigation.dart
import '../native_cpp.dart';
import 'dart:ffi'; // For FFI

/// double accCmd = proportionalNavigationConfig_.n * losRate;
final double Function(
    Pointer<Double> pCurPos,
    Pointer<Double>
        pTargetPos) pureProportionalNavigationCalcAccCmd = cppRoboticsLib
    .lookup<NativeFunction<Double Function(Pointer<Double>, Pointer<Double>)>>(
        'PureProportionalNavigationCalcAccCmd')
    .asFunction();

/// double accCmd = proportionalNavigationConfig_.n * losRate;
final double Function(double losRate)
    pureProportionalNavigationCalcAccCmdFromLosRate = cppRoboticsLib
        .lookup<NativeFunction<Double Function(Double)>>(
            'PureProportionalNavigationCalcAccCmdFromLosRate')
        .asFunction();

/// params[0] = config.n;
final Pointer<Double> Function() pureProportionalNavigationGetParams =
    cppRoboticsLib
        .lookup<NativeFunction<Pointer<Double> Function()>>(
            'PureProportionalNavigationGetParams')
        .asFunction();

final void Function(double n) pureProportionalNavigationSetParams =
    cppRoboticsLib
        .lookup<NativeFunction<Void Function(Double)>>(
            'PureProportionalNavigationSetParams')
        .asFunction();

final double Function(
    Pointer<Double> pCurPos,
    Pointer<Double>
        pTargetPos) pureProportionalNavigationSetPrevValues = cppRoboticsLib
    .lookup<NativeFunction<Double Function(Pointer<Double>, Pointer<Double>)>>(
        'PureProportionalNavigationSetPrevValues')
    .asFunction();

File: cppRobotics/extern/guidance/PureProportionalNavigation.cpp
#include "PureProportionalNavigation.h"

static ProportionalNavigation proportionalNavigation(50);
static double pParams[1];

double PureProportionalNavigationCalcAccCmd(double* pCurPos, double* pTargetPos) {
  Eigen::Vector3d curPos = {pCurPos[0], pCurPos[1], pCurPos[2]};
  Eigen::Vector3d targetPos = {pTargetPos[0], pTargetPos[1], pTargetPos[2]};
  PureProportionalNavigationData curInputData = {curPos, targetPos};
  return proportionalNavigation.CalcAccCmd(curInputData);
}

double PureProportionalNavigationCalcAccCmdFromLosRate(double losRate) {
  return proportionalNavigation.CalcAccCmdFromLosRate(losRate);
}

double* PureProportionalNavigationGetParams() {
  ProportionalNavigationConfig config = proportionalNavigation.GetParams();
  pParams[0] = config.n;
  return pParams;
}

void PureProportionalNavigationSetParams(double n) {
  ProportionalNavigationConfig config{n};
  proportionalNavigation.SetParams(config);
}

void PureProportionalNavigationSetPrevValues(double* pCurPos, double* pTargetPos) {
  Eigen::Vector3d curPos = {pCurPos[0], pCurPos[1], pCurPos[2]};
  Eigen::Vector3d targetPos = {pTargetPos[0], pTargetPos[1], pTargetPos[2]};
  PureProportionalNavigationData prevInputData = {curPos, targetPos};
  proportionalNavigation.SetPrevData(prevInputData);
}
